### PR TITLE
Rewrite culling to be more cache/thread friendly.

### PR DESCRIPTION
**Pull request by reduz**
-Uses a single array with all data.
-Massive performance improvement (Ryzen 7, 10k objects, culling goes down from 4ms to 1.2ms).
-Does not support threads yet, but code is now thread friendly, so this should improve performance several times more eventually.

### CODE REVIEW

File: servers/rendering/renderer_rd/renderer_scene_render_forward.cpp [hunk 1]
```diff
@@ -1532,21 +1532,28 @@ void RendererSceneRenderForward::_add_geometry_with_material(InstanceBase *p_ins
...
+	Plane near_plane(p_cam_transform.origin, -p_cam_transform.basis.get_axis(Vector3::AXIS_Z));
+	near_plane.d += p_cam_projection.get_z_near();
+	float z_max = p_cam_projection.get_z_far() - p_cam_projection.get_z_near();
+
 	uint32_t geometry_index = 0;
 
 	//fill list
 
 	for (int i = 0; i < (int)p_instances.size(); i++) {
 		InstanceBase *inst = p_instances[i];
```
Comment (reduz) [inline servers/rendering/renderer_rd/renderer_scene_render_forward.cpp:1553]: This code was moved here from cull. I am not sure why it was in there originally, as it makes more sense here.

File: servers/rendering/renderer_scene_cull.cpp [hunk 2]
```diff
@@ -1322,423 +1488,311 @@ void RendererSceneCull::_update_instance_lightmap_captures(Instance *p_instance)
...
+		//obtain the frustum endpoints
 
-				// setup a camera matrix for that range!
-				CameraMatrix camera_matrix;
+		Vector3 endpoints[8]; // frustum plane endpoints
+		bool res = camera_matrix.get_endpoints(p_cam_transform, endpoints);
+		ERR_CONTINUE(!res);
 
-				real_t aspect = p_cam_projection.get_aspect();
+		// obtain the light frustm ranges (given endpoints)
```
Comment (clayjohn) [inline servers/rendering/renderer_scene_cull.cpp:1568]: typo

File: servers/rendering/renderer_scene_cull.cpp [hunk 3]
```diff
@@ -1322,423 +1488,311 @@ void RendererSceneCull::_update_instance_lightmap_captures(Instance *p_instance)
...
 
-					camera_matrix.set_orthogonal(vp_he.y * 2.0, aspect, distances[(i == 0 || !overlap) ? i : i - 1], distances[i + 1], false);
-				} else {
-					real_t fov = p_cam_projection.get_fov(); //this is actually yfov, because set aspect tries to keep it
-					camera_matrix.set_perspective(fov, aspect, distances[(i == 0 || !overlap) ? i : i - 1], distances[i + 1], true);
-				}
+		Vector3 x_vec = transform.basis.get_axis(Vector3::AXIS_X).normalized();
+		Vector3 y_vec = transform.basis.get_axis(Vector3::AXIS_Y).normalized();
+		Vector3 z_vec = transform.basis.get_axis(Vector3::AXIS_Z).normalized();
+		//z_vec points agsint the camera, like in default opengl
```
Comment (clayjohn) [inline servers/rendering/renderer_scene_cull.cpp:1575]: typo

File: servers/rendering/renderer_scene_cull.cpp [hunk 4]
```diff
@@ -1322,423 +1488,311 @@ void RendererSceneCull::_update_instance_lightmap_captures(Instance *p_instance)
...
-				p_scenario->indexers[Scenario::INDEXER_GEOMETRY].convex_query(light_frustum_planes.ptr(), light_frustum_planes.size(), points.ptr(), points.size(), cull_convex);
+				x_max_cam = Math::stepify(x_max_cam, unit);
+				x_min_cam = Math::stepify(x_min_cam, unit);
+				y_max_cam = Math::stepify(y_max_cam, unit);
+				y_min_cam = Math::stepify(y_min_cam, unit);
+			}
+		}
 
-				// a pre pass will need to be needed to determine the actual z-near to be used
+		//now that we now all ranges, we can proceed to make the light frustum planes, for culling octree
```
Comment (clayjohn) [inline servers/rendering/renderer_scene_cull.cpp:1684]: typo - now that we know

File: servers/rendering/renderer_scene_cull.cpp [hunk 5]
```diff
@@ -856,6 +931,23 @@ void RendererSceneCull::instance_geometry_set_cast_shadows_setting(RID p_instanc
...
+		if (instance->cast_shadows != RS::SHADOW_CASTING_SETTING_SHADOWS_ONLY) {
+			idata.flags |= InstanceData::FLAG_CAST_SHADOWS;
+		} else {
+			idata.flags &= ~uint32_t(InstanceData::FLAG_CAST_SHADOWS);
+		}
+
+		if (instance->cast_shadows == RS::SHADOW_CASTING_SETTING_SHADOWS_ONLY) {
+			idata.flags |= InstanceData::FLAG_CAST_SHADOWS_ONLY;
+		} else {
+			idata.flags &= ~uint32_t(InstanceData::FLAG_CAST_SHADOWS_ONLY);
```
[the remark this answers is not in the document]
Comment (reduz) [reply servers/rendering/renderer_scene_cull.cpp:947]: yeah its ok, the others make no difference for the culling

File: servers/rendering/renderer_scene_cull.cpp [hunk 6]
```diff
@@ -1091,12 +1187,60 @@ void RendererSceneCull::_update_instance(Instance *p_instance) {
...
+			default: {
+			}
+		}
+
+		if (p_instance->base_type == RS::INSTANCE_REFLECTION_PROBE) {
+			//always dirty when added
+			idata.flags |= InstanceData::FLAG_REFLECTION_PROBE_DIRTY;
+		}
+		if (p_instance->cast_shadows != RS::SHADOW_CASTING_SETTING_SHADOWS_ONLY) {
+			idata.flags |= InstanceData::FLAG_CAST_SHADOWS;
```
[the remark this answers is not in the document]
Comment (reduz) [reply servers/rendering/renderer_scene_cull.cpp:1219]: again, no difference for culling

**Review by clayjohn — APPROVED** [review]

**Pull request merged**